ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,436 @@
 
 #include "bcmath.h"
 #include "private.h"
+#include "convert.h"
 #include <stdbool.h>
 #include <stddef.h>
 #include <string.h>
 #include "zend_alloc.h"
 
+static const BC_VECTOR POW_10_LUT[9] = {
+	1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000
+};
 
-/* Some utility routines for the divide:  First a one digit multiply.
-   NUM (with SIZE digits) is multiplied by DIGIT and the result is
-   placed into RESULT.  It is written so that NUM and RESULT can be
-   the same pointers.  */
-
-static void _one_mult(unsigned char *num, size_t size, int digit, unsigned char *result)
+/*
+ * Use this function when the number of array elements in n2 is 1, that is, when n2 is not divided into multiple array elements.
+ * The algorithm can be simplified if n2 is not divided, and this function is an optimization of bc_standard_div.
+ */
+static inline void bc_fast_div(BC_VECTOR *n1_vector, size_t n1_arr_size, BC_VECTOR n2_vector, BC_VECTOR *quot_vector, size_t quot_arr_size)
 {
-	size_t carry, value;
-	unsigned char *nptr, *rptr;
+	size_t n1_top_index = n1_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+	for (size_t i = 0; i < quot_arr_size - 1; i++) {
+		if (n1_vector[n1_top_index - i] < n2_vector) {
+			quot_vector[quot_top_index - i] = 0;
+			/* n1_vector[n1_top_index - i] is always a smaller number than n2_vector. Therefore, there will be no overflow. */
+			n1_vector[n1_top_index - i - 1] += n1_vector[n1_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+			continue;
+		}
+		quot_vector[quot_top_index - i] = n1_vector[n1_top_index - i] / n2_vector;
+		n1_vector[n1_top_index - i] -= n2_vector * quot_vector[quot_top_index - i];
+		n1_vector[n1_top_index - i - 1] += n1_vector[n1_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+		n1_vector[n1_top_index - i] = 0;
+	}
+	/* last */
+	quot_vector[0] = n1_vector[0] / n2_vector;
+}
 
-	if (digit == 0) {
-		memset(result, 0, size);
-	} else {
-		if (digit == 1) {
-			memcpy(result, num, size);
-		} else {
-			/* Initialize */
-			nptr = (unsigned char *) (num + size - 1);
-			rptr = (unsigned char *) (result + size - 1);
-			carry = 0;
-
-			while (size-- > 0) {
-				value = *nptr-- * digit + carry;
-				*rptr-- = value % BASE;
-				carry = value / BASE;
-			}
+/*
+ * Used when the number of elements in the n2 vector is 2 or more.
+ */
+static inline void bc_standard_div(
+	BC_VECTOR *n1_vector, size_t n1_arr_size,
+	BC_VECTOR *n2_vector, size_t n2_arr_size, size_t n2_len,
```

Comment:
```suggestion
	BC_VECTOR *n2_vector, size_t n2_arr_size
```
Seems like a duplicate param?

Also maybe name the size param `nd_vector_size` instead of `arr`?

Reply:
No, n2_arr_size means the size of the vector array, and n2_len means the number of digits in n2.

e.g.
n2 = 1234567890 (64-bit, vector size = 8)
n2_arr_size = 2
n2_len = 10